File: skills/clawhub/skills-skip/enshrined-mint/mint.py
#!/usr/bin/env python3
"""
Enshrined Exchange - Mint/Redeem/Swap/Check Balance
Network: Tempo (Chain 4217)
RPC: https://rpc.moderato.tempo.xyz

Usage:
  Mint:    PRIVATE_KEY=0x... python mint.py mint --amount 100 --source USDC
  Redeem:  PRIVATE_KEY=0x... python mint.py redeem --amount 100 --dest USDC
  Swap:    PRIVATE_KEY=0x... python mint.py swap --amount 100 --from USDC --to USDe
  Balance: PRIVATE_KEY=0x... python mint.py balance

Requirements:
  pip install web3
"""

import os
import sys
import argparse
from decimal import Decimal

from web3 import Web3

RPC = 'https://rpc.moderato.tempo.xyz'
EXCHANGE = Web3.to_checksum_address('0xdec0000000000000000000000000000000000000')

TOKENS = {
    'USDC':   Web3.to_checksum_address('0x20C0000000000000000000000000000000000000'),
    'USDE':    Web3.to_checksum_address('0x20C0000000000000000000000000000000000001'),
    'PATHUSD': Web3.to_checksum_address('0x20C0000000000000000000000000000000000002'),
}

DECIMALS = {'USDC': 6, 'USDE': 18, 'PATHUSD': 18}

EXPLORER = 'https://explorer.tempo.xyz/tx'

EXCHANGE_ABI = [
    {
        'inputs': [
            {'name': 'tokenIn', 'type': 'address'},
            {'name': 'tokenOut', 'type': 'address'},
            {'name': 'amount', 'type': 'uint128'},
            {'name': 'minAmountOut', 'type': 'uint128'},
        ],
        'name': 'swapExactAmountIn',
        'outputs': [{'name': '', 'type': 'uint128'}],
        'stateMutability': 'nonpayable',
        'type': 'function',
    },
    {
        'inputs': [
            {'name': 'user', 'type': 'address'},
            {'name': 'token', 'type': 'address'},
        ],
        'name': 'getBalance',
        'outputs': [{'name': '', 'type': 'uint256'}],
        'stateMutability': 'view',
        'type': 'function',
    },
]

TOKEN_ABI = [
    {
        'inputs': [
            {'name': 'to', 'type': 'address'},
            {'name': 'amount', 'type': 'uint256'},
        ],
        'name': 'mint',
        'outputs': [{'name': '', 'type': 'bool'}],
        'stateMutability': 'nonpayable',
        'type': 'function',
    },
    {
        'inputs': [
            {'name': 'to', 'type': 'address'},
            {'name': 'amount', 'type': 'uint256'},
        ],
        'name': 'redeem',
        'outputs': [{'name': '', 'type': 'bool'}],
        'stateMutability': 'nonpayable',
        'type': 'function',
    },
    {
        'inputs': [{'name': 'account', 'type': 'address'}],
        'name': 'balanceOf',
        'outputs': [{'name': '', 'type': 'uint256'}],
        'stateMutability': 'view',
        'type': 'function',
    },
    {
        'inputs': [
            {'name': 'spender', 'type': 'address'},
            {'name': 'amount', 'type': 'uint256'},
        ],
        'name': 'approve',
        'outputs': [{'name': '', 'type': 'bool'}],
        'stateMutability': 'nonpayable',
        'type': 'function',
    },
    {
        'inputs': [
            {'name': 'owner', 'type': 'address'},
            {'name': 'spender', 'type': 'address'},
        ],
        'name': 'allowance',
        'outputs': [{'name': '', 'type': 'uint256'}],
        'stateMutability': 'view',
        'type': 'function',
    },
]


def parse_amount(amount_str, symbol):
    """Convert decimal string to smallest unit (wei/etc)."""
    dec = DECIMALS.get(symbol, 18)
    return int(Decimal(amount_str) * (10 ** dec))


def format_amount(raw_amount, symbol):
    """Convert smallest unit to decimal string."""
    dec = DECIMALS.get(symbol, 18)
    return str(Decimal(raw_amount) / (10 ** dec))


def check_balance(w3, token_addr, wallet_addr, symbol):
    """Get wallet + exchange balance for a token."""
    token = w3.eth.contract(address=token_addr, abi=TOKEN_ABI)
    exchange = w3.eth.contract(address=EXCHANGE, abi=EXCHANGE_ABI)

    wallet_bal = token.functions.balanceOf(wallet_addr).call()
    try:
        exchange_bal = exchange.functions.getBalance(wallet_addr, token_addr).call()
    except Exception:
        exchange_bal = 0

    print(f'  {symbol} (wallet):   {format_amount(wallet_bal, symbol)}')
    print(f'  {symbol} (exchange): {format_amount(exchange_bal, symbol)}')
    return wallet_bal


def action_balance(w3, wallet_addr, token_symbol=None):
    """Show balance for one or all tokens."""
    tokens = [token_symbol.upper()] if token_symbol else TOKENS.keys()
    print('\n📊 Token Balances')
    for t in tokens:
        if t not in TOKENS:
            print(f'  Unknown token: {t}')
            continue
        check_balance(w3, TOKENS[t], wallet_addr, t)


def action_mint(w3, wallet, amount_str, source_token):
    """Mint USDe by depositing source stablecoin."""
    source = source_token.upper()
    if source not in TOKENS:
        raise ValueError(f'Unknown source token: {source}')
    if source == 'USDE':
        raise ValueError('Source must be a stablecoin (USDC, pathUSD), not USDe')

    source_addr = TOKENS[source]
    usde_addr = TOKENS['USDE']
    amount = parse_amount(amount_str, source)

    token_src = w3.eth.contract(address=source_addr, abi=TOKEN_ABI)
    token_usde = w3.eth.contract(address=usde_addr, abi=TOKEN_ABI)

    # Check source balance
    src_bal = token_src.functions.balanceOf(wallet.address).call()
    print(f'\n📋 Step 1: {source} balance check')
    print(f'  {source} balance: {format_amount(src_bal, source)}')
    if src_bal < amount:
        raise ValueError(
            f'Insufficient {source} balance. '
            f'Have {format_amount(src_bal, source)}, need {format_amount(amount, source)}'
        )

    # Approve USDe contract
    current_allow = token_src.functions.allowance(wallet.address, usde_addr).call()
    print(f'\n📋 Step 2: Approve USDe contract to spend {source}')
    if current_allow < amount:
        tx = token_src.functions.approve(usde_addr, 2**256 - 1).build_transaction({
            'from': wallet.address,
            'nonce': w3.eth.get_transaction_count(wallet.address),
            'gas': 100000,
            'gasPrice': w3.eth.gas_price,
        })
        signed = wallet.sign_transaction(tx)
        tx_hash = w3.eth.send_raw_transaction(signed.rawTransaction)
        receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
        print(f'  ✅ Approval confirmed ({receipt["transactionHash"].hex()})')
    else:
        print('  ✅ Already approved')

    # Mint
    print(f'\n📋 Step 3: Mint {amount_str} USDe')
    tx = token_usde.functions.mint(wallet.address, amount).build_transaction({
        'from': wallet.address,
        'nonce': w3.eth.get_transaction_count(wallet.address),
        'gas': 200000,
        'gasPrice': w3.eth.gas_price,
    })
    signed = wallet.sign_transaction(tx)
    tx_hash = w3.eth.send_raw_transaction(signed.rawTransaction)
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)

    print(f'\n✅ MINT SUCCESSFUL')
    print(f'  TX: {EXPLORER}/{tx_hash.hex()}')
    print(f'  Gas used: {receipt["gasUsed"]}')

    # Final balances
    final_src = token_src.functions.balanceOf(wallet.address).call()
    final_usde = token_usde.functions.balanceOf(wallet.address).call()
    print(f'\n📊 Final Balances')
    print(f'  {source}: {format_amount(final_src, source)}')
    print(f'  USDe: {format_amount(final_usde, "USDe")}')


def action_redeem(w3, wallet, amount_str, dest_token):
    """Redeem USDe for destination stablecoin."""
    dest = dest_token.upper()
    if dest not in TOKENS:
        raise ValueError(f'Unknown dest token: {dest}')
    if dest == 'USDE':
        raise ValueError('Dest must be a stablecoin (USDC, pathUSD), not USDe')

    dest_addr = TOKENS[dest]
    usde_addr = TOKENS['USDE']
    amount = parse_amount(amount_str, 'USDE')

    token_usde = w3.eth.contract(address=usde_addr, abi=TOKEN_ABI)
    token_dest = w3.eth.contract(address=dest_addr, abi=TOKEN_ABI)

    # Check USDe balance
    usde_bal = token_usde.functions.balanceOf(wallet.address).call()
    print(f'\n📋 Step 1: USDe balance check')
    print(f'  USDe balance: {format_amount(usde_bal, "USDe")}')
    if usde_bal < amount:
        raise ValueError(
            f'Insufficient USDe balance. '
            f'Have {format_amount(usde_bal, "USDe")}, need {format_amount(amount, "USDe")}'
        )

    # Redeem
    print(f'\n📋 Step 2: Redeem {amount_str} USDe → {dest}')
    tx = token_usde.functions.redeem(wallet.address, amount).build_transaction({
        'from': wallet.address,
        'nonce': w3.eth.get_transaction_count(wallet.address),
        'gas': 200000,
        'gasPrice': w3.eth.gas_price,
    })
    signed = wallet.sign_transaction(tx)
    tx_hash = w3.eth.send_raw_transaction(signed.rawTransaction)
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)

    print(f'\n✅ REDEEM SUCCESSFUL')
    print(f'  TX: {EXPLORER}/{tx_hash.hex()}')
    print(f'  Gas used: {receipt["gasUsed"]}')

    # Final balances
    final_usde = token_usde.functions.balanceOf(wallet.address).call()
    final_dest = token_dest.functions.balanceOf(wallet.address).call()
    print(f'\n📊 Final Balances')
    print(f'  USDe: {format_amount(final_usde, "USDe")}')
    print(f'  {dest}: {format_amount(final_dest, dest)}')


def action_swap(w3, wallet, amount_str, from_token, to_token):
    """Swap one stablecoin for another via exchange."""
    src = from_token.upper()
    dst = to_token.upper()

    if src not in TOKENS:
        raise ValueError(f'Unknown from token: {src}')
    if dst not in TOKENS:
        raise ValueError(f'Unknown to token: {dst}')
    if src == dst:
        raise ValueError('From and To cannot be the same')

    src_addr = TOKENS[src]
    dst_addr = TOKENS[dst]
    amount = parse_amount(amount_str, src)

    exchange = w3.eth.contract(address=EXCHANGE, abi=EXCHANGE_ABI)
    token_src = w3.eth.contract(address=src_addr, abi=TOKEN_ABI)

    # Check source balance
    src_bal = token_src.functions.balanceOf(wallet.address).call()
    print(f'\n📋 Step 1: {src} balance check')
    print(f'  {src} balance: {format_amount(src_bal, src)}')
    if src_bal < amount:
        raise ValueError(f'Insufficient {src} balance')

    # Approve exchange
    current_allow = token_src.functions.allowance(wallet.address, EXCHANGE).call()
    print(f'\n📋 Step 2: Approve exchange to spend {src}')
    if current_allow < amount:
        tx = token_src.functions.approve(EXCHANGE, 2**256 - 1).build_transaction({
            'from': wallet.address,
            'nonce': w3.eth.get_transaction_count(wallet.address),
            'gas': 100000,
            'gasPrice': w3.eth.gas_price,
        })
        signed = wallet.sign_transaction(tx)
        tx_hash = w3.eth.send_raw_transaction(signed.rawTransaction)
        receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
        print(f'  ✅ Approval confirmed ({receipt["transactionHash"].hex()})')
    else:
        print('  ✅ Already approved')

    # Swap
    print(f'\n📋 Step 3: Swap {amount_str} {src} → {dst}')
    tx = exchange.functions.swapExactAmountIn(src_addr, dst_addr, amount, 0).build_transaction({
        'from': wallet.address,
        'nonce': w3.eth.get_transaction_count(wallet.address),
        'gas': 300000,
        'gasPrice': w3.eth.gas_price,
    })
    signed = wallet.sign_transaction(tx)
    tx_hash = w3.eth.send_raw_transaction(signed.rawTransaction)
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)

    print(f'\n✅ SWAP SUCCESSFUL')
    print(f'  TX: {EXPLORER}/{tx_hash.hex()}')
    print(f'  Gas used: {receipt["gasUsed"]}')

    # Final balances
    final_src = token_src.functions.balanceOf(wallet.address).call()
    final_dst_bal = exchange.functions.getBalance(wallet.address, dst_addr).call()
    print(f'\n📊 Final Balances')
    print(f'  {src}: {format_amount(final_src, src)}')
    print(f'  {dst} (exchange): {format_amount(final_dst_bal, dst)}')


def main():
    parser = argparse.ArgumentParser(description='Enshrined Exchange - Mint/Redeem/Swap on Tempo')
    parser.add_argument('action', choices=['mint', 'redeem', 'swap', 'balance'],
                        help='Action to perform')
    parser.add_argument('--amount', help='Amount (e.g. 100)')
    parser.add_argument('--source', default='USDC',
                        help='Source token for minting (default: USDC)')
    parser.add_argument('--dest', default='USDC',
                        help='Destination token for redeem (default: USDC)')
    parser.add_argument('--from', dest='from_token', default='USDC',
                        help='From token for swap (default: USDC)')
    parser.add_argument('--to', dest='to_token', default='USDE',
                        help='To token for swap (default: USDe)')
    parser.add_argument('--token', dest='token_symbol',
                        help='Specific token for balance check')

    args = parser.parse_args(sys.argv[1:])

    private_key = os.environ.get('PRIVATE_KEY')
    if not private_key:
        print('❌ PRIVATE_KEY not set\n', file=sys.stderr)
        print('Usage:', file=sys.stderr)
        print('  Mint:    PRIVATE_KEY=0x... python mint.py mint --amount 100 --source USDC', file=sys.stderr)
        print('  Redeem:  PRIVATE_KEY=0x... python mint.py redeem --amount 100 --dest USDC', file=sys.stderr)
        print('  Swap:    PRIVATE_KEY=0x... python mint.py swap --amount 100 --from USDC --to USDe', file=sys.stderr)
        print('  Balance: PRIVATE_KEY=0x... python mint.py balance', file=sys.stderr)
        print('\nTokens: USDC, USDe, pathUSD', file=sys.stderr)
        sys.exit(1)

    w3 = Web3(Web3.HTTPProvider(RPC))
    if not w3.is_connected():
        print('❌ Cannot connect to RPC', file=sys.stderr)
        sys.exit(1)

    wallet = w3.eth.account.from_key(private_key)

    print('')
    print('🔷 ENSHRIINED.EXCHANGE')
    print('═══════════════════════════════════════════════')
    print(f'  Network:  Tempo (Chain 4217)')
    print(f'  RPC:      {RPC}')
    print(f'  Wallet:   {wallet.address}')
    print(f'  Action:   {args.action}')

    try:
        if args.action == 'balance':
            action_balance(w3, wallet.address, args.token_symbol)
        elif args.action == 'mint':
            if not args.amount:
                raise ValueError('--amount required for mint')
            action_mint(w3, wallet, args.amount, args.source)
        elif args.action == 'redeem':
            if not args.amount:
                raise ValueError('--amount required for redeem')
            action_redeem(w3, wallet, args.amount, args.dest)
        elif args.action == 'swap':
            if not args.amount:
                raise ValueError('--amount required for swap')
            action_swap(w3, wallet, args.amount, args.from_token, args.to_token)
    except Exception as err:
        print(f'\n❌ Error: {err}', file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()

File: skills/clawhub/skills-skip/enshrined-mint/mint.js
#!/usr/bin/env node
/**
 * Enshrined Exchange - Mint/Swap/Check Balance (Tempo DEX)
 *
 * Network:  Tempo (Chain 4217)
 * RPC:      https://rpc.moderato.tempo.xyz
 * Explorer: https://explorer.tempo.xyz
 *
 * Key Contracts:
 *   Stablecoin DEX:  0xdec0000000000000000000000000000000000000
 *   pathUSD:         0x20c0000000000000000000000000000000000000
 *   ENSH:            0x20c00000000000000000000031d99efa5dbd3713
 *   ENSH Mint:       0x3ba0a6d270f7dd69964f67219288f8cedfab54bc
 *
 * Usage:
 *   PRIVATE_KEY=0x... node mint.js --action swap --amount 3600 --from pathUSD --to ENSH
 *   PRIVATE_KEY=0x... node mint.js --action balance
 *   PRIVATE_KEY=0x... node mint.js --action mint-ensh --amount 100 --source pathUSD
 *
 * Docs: https://docs.tempo.xyz/protocol/exchange/spec
 */

const { ethers } = require('ethers');
const args = require('minimist')(process.argv.slice(2));

const RPC = 'https://rpc.moderato.tempo.xyz';

// Contracts
const DEX = '0xdec0000000000000000000000000000000000000';
const ENSH_MINT = '0x3ba0a6d270f7dd69964f67219288f8cedfab54bc';

const TOKENS = {
  pathUSD: '0x20c0000000000000000000000000000000000000',
  ENSH:    '0x20c00000000000000000031d99efa5dbd3713',
};

const DECIMALS = { pathUSD: 6, ENSH: 18 };
const EXPLORER = 'https://explorer.tempo.xyz/tx';

// ─── Stablecoin DEX ABI (from docs.tempo.xyz) ─────────────────────────────────
const DEX_ABI = [
  // Constants
  'function PRICE_SCALE() external view returns (uint32)',
  'function TICK_SPACING() external view returns (int16)',
  'function MIN_TICK() external view returns (int16)',
  'function MAX_TICK() external view returns (int16)',
  // Balances
  'function balanceOf(address user, address token) external view returns (uint128)',
  'function withdraw(address token, uint128 amount) external',
  // Swaps
  'function swapExactAmountIn(address tokenIn, address tokenOut, uint128 amountIn, uint128 minAmountOut) external returns (uint128 amountOut)',
  'function swapExactAmountOut(address tokenIn, address tokenOut, uint128 amountOut, uint128 maxAmountIn) external returns (uint128 amountIn)',
  'function quoteSwapExactAmountIn(address tokenIn, address tokenOut, uint128 amountIn) external view returns (uint128 amountOut)',
  'function quoteSwapExactAmountOut(address tokenIn, address tokenOut, uint128 amountOut) external view returns (uint128 amountIn)',
  // Orders
  'function place(address token, uint128 amount, bool isBid, int16 tick) external returns (uint128 orderId)',
  'function cancel(uint128 orderId) external',
  // Books
  'function pairKey(address tokenA, address tokenB) external pure returns (bytes32 key)',
  'function books(bytes32 pairKey) external view returns (address base, address quote, int16 bestBidTick, int16 bestAskTick)',
];

// ERC-20 ABI
const ERC20_ABI = [
  'function balanceOf(address) view returns (uint256)',
  'function decimals() view returns (uint8)',
  'function symbol() view returns (string)',
  'function approve(address spender, uint256 amount) returns (bool)',
  'function allowance(address owner, address spender) view returns (uint256)',
  'function transfer(address to, uint256 amount) returns (bool)',
  'function transferFrom(address from, address to, uint256 amount) returns (bool)',
];

// ENSH Mint contract ABI (unknown - probing)
const MINT_ABI = [
  'function mint(address to, uint256 amount)',
  'function deposit(uint256 amount)',
  'function swap(uint256 amount)',
  'function enter(address token, uint256 amount)',
  'function buy(address token, uint256 amount)',
];

function parseAmount(amount, symbol) {
  const dec = DECIMALS[symbol] || 18;
  return ethers.parseUnits(amount.toString(), dec);
}

function formatAmount(amount, symbol) {
  const dec = DECIMALS[symbol] || 18;
  return ethers.formatUnits(amount, dec);
}

async function getBalance(contract, address) {
  try {
    return await contract.balanceOf(address);
  } catch { return 0n; }
}

async function actionBalance(wallet, tokenSymbol) {
  const tokenAddr = TOKENS[tokenSymbol];
  if (!tokenAddr) throw new Error(`Unknown token: ${tokenSymbol}`);

  // ERC20 wallet balance
  const token = new ethers.Contract(tokenAddr, ERC20_ABI, wallet);
  const bal = await getBalance(token, wallet.address);
  console.log(`  ${tokenSymbol} (wallet): ${formatAmount(bal, tokenSymbol)}`);

  // DEX internal balance
  const dex = new ethers.Contract(DEX, DEX_ABI, wallet);
  try {
    const dexBal = await dex.balanceOf(wallet.address, tokenAddr);
    console.log(`  ${tokenSymbol} (DEX):   ${formatAmount(dexBal, tokenSymbol)}`);
  } catch { /* skip */ }
}

async function actionSwap(wallet, amount, fromToken, toToken) {
  const src = fromToken.toUpperCase();
  const dst = toToken.toUpperCase();
  const srcAddr = TOKENS[fromToken];
  const dstAddr = TOKENS[toToken];

  if (!srcAddr) throw new Error(`Unknown token: ${fromToken}`);
  if (!dstAddr) throw new Error(`Unknown token: ${toToken}`);
  if (src === dst) throw new Error('From and To cannot be the same');

  const amountIn = parseAmount(amount, fromToken);
  const dex = new ethers.Contract(DEX, DEX_ABI, wallet);
  const token = new ethers.Contract(srcAddr, ERC20_ABI, wallet);

  // Step 1: Check wallet balance
  console.log(`\n📋 Step 1: Check ${fromToken} balance`);
  const walletBal = await getBalance(token, wallet.address);
  console.log(`  Wallet: ${formatAmount(walletBal, fromToken)}`);
  if (walletBal < amountIn) throw new Error(`Insufficient ${fromToken} balance`);

  // Step 2: Check DEX internal balance
  console.log(`\n📋 Step 2: Check DEX internal balance`);
  try {
    const dexBal = await dex.balanceOf(wallet.address, srcAddr);
    console.log(`  DEX internal: ${formatAmount(dexBal, fromToken)}`);
    console.log(`  Note: DEX uses internal balances — use withdraw() to pull funds out`);
  } catch { /* skip */ }

  // Step 3: Approve DEX to spend token
  console.log(`\n📋 Step 3: Approve DEX to spend ${fromToken}`);
  const currentAllow = await token.allowance(wallet.address, DEX).catch(() => 0n);
  if (currentAllow < amountIn) {
    const approveTx = await token.approve(DEX, ethers.MaxUint256);
    await approveTx.wait();
    console.log('  ✅ Approved');
  } else {
    console.log('  ✅ Already approved');
  }

  // Step 4: Get a quote first
  console.log(`\n📋 Step 4: Quote`);
  try {
    const quote = await dex.quoteSwapExactAmountIn(srcAddr, dstAddr, amountIn);
    console.log(`  Expected output: ${formatAmount(quote, toToken)} ${toToken}`);
  } catch(e) {
    console.log(`  ⚠️  Quote failed: ${e.message.slice(0,80)}`);
  }

  // Step 5: Execute swap
  console.log(`\n📋 Step 5: Swap ${amount} ${fromToken} → ${toToken}`);
  try {
    const swapTx = await dex.swapExactAmountIn(srcAddr, dstAddr, amountIn, 0n);
    const receipt = await swapTx.wait();
    console.log(`\n✅ SWAP SUCCESSFUL`);
    console.log(`  TX: ${EXPLORER}/${swapTx.hash}`);
    console.log(`  Gas used: ${receipt.gasUsed.toString()}`);
  } catch(e) {
    // Try swapExactAmountOut as fallback
    console.log(`  swapExactAmountIn failed, trying swapExactAmountOut...`);
    const swapTx = await dex.swapExactAmountOut(srcAddr, dstAddr, amountIn, ethers.MaxUint256);
    const receipt = await swapTx.wait();
    console.log(`\n✅ SWAP SUCCESSFUL`);
    console.log(`  TX: ${EXPLORER}/${swapTx.hash}`);
  }

  // Final balances
  const finalSrc = await getBalance(token, wallet.address);
  const dexDst = await dex.balanceOf(wallet.address, dstAddr).catch(() => 0n);
  console.log(`\n📊 Final`);
  console.log(`  ${fromToken} (wallet): ${formatAmount(finalSrc, fromToken)}`);
  console.log(`  ${toToken} (DEX):     ${formatAmount(dexDst, toToken)}`);
}

async function actionMintENSH(wallet, amount, sourceToken) {
  const src = sourceToken.toUpperCase();
  const srcAddr = TOKENS[sourceToken];
  if (!srcAddr) throw new Error(`Unknown token: ${sourceToken}`);
  if (src === 'ENSH') throw new Error('Source must be a stablecoin for minting');

  const amountIn = parseAmount(amount, sourceToken);
  const token = new ethers.Contract(srcAddr, ERC20_ABI, wallet);
  const mint = new ethers.Contract(ENSH_MINT, MINT_ABI, wallet);

  // Check balance
  console.log(`\n📋 Step 1: Check ${sourceToken} balance`);
  const bal = await getBalance(token, wallet.address);
  console.log(`  Balance: ${formatAmount(bal, sourceToken)}`);
  if (bal < amountIn) throw new Error(`Insufficient ${sourceToken}`);

  // Approve mint contract
  console.log(`\n📋 Step 2: Approve mint contract`);
  const currentAllow = await token.allowance(wallet.address, ENSH_MINT).catch(() => 0n);
  if (currentAllow < amountIn) {
    const approveTx = await token.approve(ENSH_MINT, ethers.MaxUint256);
    await approveTx.wait();
    console.log('  ✅ Approved');
  } else {
    console.log('  ✅ Already approved');
  }

  // Try mint functions
  console.log(`\n📋 Step 3: Mint ENSH`);

  // Try each mint function
  const mintFns = [
    { name: 'mint(address,uint256)', fn: 'mint', args: [wallet.address, amountIn] },
    { name: 'swap(uint256)', fn: 'swap', args: [amountIn] },
    { name: 'enter(address,uint256)', fn: 'enter', args: [srcAddr, amountIn] },
    { name: 'buy(address,uint256)', fn: 'buy', args: [srcAddr, amountIn] },
    { name: 'deposit(uint256)', fn: 'deposit', args: [amountIn] },
  ];

  let success = false;
  for (const mintFn of mintFns) {
    try {
      console.log(`  Trying ${mintFn.name}...`);
      const tx = await mint[mintFn.fn](...mintFn.args);
      const receipt = await tx.wait();
      console.log(`\n✅ MINT SUCCESSFUL (${mintFn.name})`);
      console.log(`  TX: ${EXPLORER}/${tx.hash}`);
      success = true;
      break;
    } catch(e) {
      if (e.message.includes('execution reverted')) {
        console.log(`    ✗ ${mintFn.name} reverted`);
      } else {
        console.log(`    ✗ ${mintFn.name} error: ${e.message.slice(0,60)}`);
      }
    }
  }

  if (!success) {
    console.log('\n⚠️  Mint functions not confirmed. Check explorer for correct ABI.');
  }
}

async function main() {
  const privateKey = process.env.PRIVATE_KEY;
  const action = (args.action || 'balance').toLowerCase();
  const amount = parseFloat(args.amount) || 0;
  const fromToken = args.from || 'pathUSD';
  const toToken = args.to || 'ENSH';
  const sourceToken = args.source || 'pathUSD';

  if (!privateKey) {
    console.error('❌ PRIVATE_KEY not set\n');
    console.error('Usage:');
    console.error('  Swap:    PRIVATE_KEY=0x... node mint.js --action swap --amount 3600 --from pathUSD --to ENSH');
    console.error('  Balance: PRIVATE_KEY=0x... node mint.js --action balance');
    console.error('  Mint:    PRIVATE_KEY=0x... node mint.js --action mint-ensh --amount 100 --source pathUSD');
    console.error('\nTokens: pathUSD, ENSH');
    process.exit(1);
  }

  const provider = new ethers.JsonRpcProvider(RPC);
  const wallet = new ethers.Wallet(privateKey, provider);

  console.log('');
  console.log('🔷 ENSHRIINED.EXCHANGE');
  console.log('═══════════════════════════════════════════════');
  console.log(`  Network:  Tempo (Chain 4217)`);
  console.log(`  RPC:      ${RPC}`);
  console.log(`  Wallet:   ${wallet.address}`);
  console.log(`  Action:   ${action}`);

  try {
    if (action === 'balance') {
      console.log('\n📊 Token Balances');
      for (const t of Object.keys(TOKENS)) {
        await actionBalance(wallet, t);
      }
    } else if (action === 'swap') {
      if (!amount) throw new Error('--amount required');
      await actionSwap(wallet, amount, fromToken, toToken);
    } else if (action === 'mint-ensh') {
      if (!amount) throw new Error('--amount required');
      await actionMintENSH(wallet, amount, sourceToken);
    } else {
      throw new Error(`Unknown action: ${action}`);
    }
  } catch (err) {
    console.error(`\n❌ Error: ${err.message || err}`);
    process.exit(1);
  }
}

main();
